[pkg/stanza] Flaky test TestDeleteAfterRead_SkipPartials

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6562693797/job/17826173436?pr=27825

```
--- FAIL: TestDeleteAfterRead_SkipPartials (10.36s)
    file_test.go:1627: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/file_test.go:1627
        	Error:      	file "/tmp/TestDeleteAfterRead_SkipPartials2080466458/001/1867497531" exists
        	Test:       	TestDeleteAfterRead_SkipPartials
FAIL
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer	coverage: [88](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6562693797/job/17826173436?pr=27825#step:9:89).6% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer	43.800s
```

Another hit on the same test
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6922266602/job/18828797459?pr=29343#step:9:97
```terminal
--- FAIL: TestDeleteAfterRead_SkipPartials (9.44s)
    file_test.go:1556: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/file_test.go:1556
        	Error:      	file "/tmp/TestDeleteAfterRead_SkipPartials3022013525/001/425630[104](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6922266602/job/18828797459?pr=29343#step:9:105)8" exists
        	Test:       	TestDeleteAfterRead_SkipPartials
FAIL
	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer	coverage: 87.7% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer	44.366s
```

--- a/pkg/stanza/fileconsumer/file_test.go
+++ b/pkg/stanza/fileconsumer/file_test.go
@@ -1494,10 +1494,8 @@ func TestReadExistingLogsWithHeader(t *testing.T) {
 }
 
 func TestDeleteAfterRead_SkipPartials(t *testing.T) {
-	bytesPerLine := 100
-	shortFileLine := tokenWithLength(bytesPerLine - 1)
+	shortFileLine := "short file line"
 	longFileLines := 100000
-	longFileFirstLine := "first line of long file\n"
 
 	require.NoError(t, featuregate.GlobalRegistry().Set(allowFileDeletion.ID(), true))
 	defer func() {
@@ -1513,15 +1511,13 @@ func TestDeleteAfterRead_SkipPartials(t *testing.T) {
 	operator.persister = testutil.NewUnscopedMockPersister()
 
 	shortFile := openTemp(t, tempDir)
-	_, err := shortFile.WriteString(string(shortFileLine) + "\n")
+	_, err := shortFile.WriteString(shortFileLine + "\n")
 	require.NoError(t, err)
 	require.NoError(t, shortFile.Close())
 
 	longFile := openTemp(t, tempDir)
-	_, err = longFile.WriteString(longFileFirstLine)
-	require.NoError(t, err)
 	for line := 0; line < longFileLines; line++ {
-		_, err := longFile.WriteString(string(tokenWithLength(bytesPerLine-1)) + "\n")
+		_, err := longFile.WriteString(string(tokenWithLength(100)) + "\n")
 		require.NoError(t, err)
 	}
 	require.NoError(t, longFile.Close())
@@ -1541,22 +1537,26 @@ func TestDeleteAfterRead_SkipPartials(t *testing.T) {
 		operator.poll(ctx)
 	}()
 
-	for !(shortOne && longOne) {
-		if line := waitForEmit(t, emitCalls); string(line.token) == string(shortFileLine) {
+	for !shortOne || !longOne {
+		if line := waitForEmit(t, emitCalls); string(line.token) == shortFileLine {
 			shortOne = true
 		} else {
 			longOne = true
 		}
 	}
 
+	// Short file was fully consumed and should eventually be deleted.
+	// Enforce assertion before canceling because EOF is not necessarily detected
+	// immediately when the token is emitted. An additional scan may be necessary.
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
+		assert.NoFileExists(c, shortFile.Name())
+	}, 100*time.Millisecond, time.Millisecond)
+
 	// Stop consuming before long file has been fully consumed
 	cancel()
 	wg.Wait()
 
-	// short file was fully consumed and should have been deleted
-	require.NoFileExists(t, shortFile.Name())
-
-	// long file was partially consumed and should NOT have been deleted
+	// Long file was partially consumed and should NOT have been deleted.
 	require.FileExists(t, longFile.Name())
 }
 

--- a/pkg/stanza/fileconsumer/internal/reader/reader.go
+++ b/pkg/stanza/fileconsumer/internal/reader/reader.go
@@ -82,7 +82,6 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 	s := scanner.New(r, r.MaxLogSize, scanner.DefaultBufferSize, r.Offset, r.splitFunc)
 
 	// Iterate over the tokenized file, emitting entries as we go
-	var eof bool
 	for {
 		select {
 		case <-ctx.Done():
@@ -92,10 +91,10 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 
 		ok := s.Scan()
 		if !ok {
-			if err := s.Error(); err == nil {
-				eof = true
-			} else {
+			if err := s.Error(); err != nil {
 				r.logger.Errorw("Failed during scan", zap.Error(err))
+			} else if r.DeleteAtEOF {
+				r.delete()
 			}
 			break
 		}
@@ -122,11 +121,7 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 				r.logger.Errorw("process: %w", zap.Error(err))
 			}
 		}
-
 		r.Offset = s.Pos()
-	}
-	if eof && r.DeleteAtEOF {
-		r.delete()
 	}
 }
 

